src/Data/Array/Polarized/Push.hs

```diff
@@ -0,0 +1,62 @@
+{-# OPTIONS_GHC -fno-warn-partial-type-signatures #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Data.Array.Polarized.Push where
+
+-- XXX: there is lots of scope for testing array functions
+
+import Data.Array.Destination (DArray)
+import qualified Data.Array.Destination as DArray
+import qualified Data.Functor.Linear as Data
+import Data.Vector (Vector)
+import qualified Data.Vector as Vector
+import Prelude.Linear hiding (zip, zipWith, foldr)
+import qualified Prelude
+
+data Array a where
+  Array :: (forall b. (a ->. b) -> DArray b ->. ()) ->. Int -> Array a
+  -- A note on implementation: `exists b. ((a -> b), DArray b)` adjoins freely
+  -- the structure of contravariant functor to `DArray`. Because it appears to
+  -- the left of an arrow, we can curry the existential quantification (and,
+  -- less crucially, the pair) so that we can spare an extra type definition.
+  --
+  -- An invariant is kept that, really, the first parameter is only passed
+  -- arrays of size @n@ (the second parameter, which is the length of the
+  -- array).
+  --
+  -- TODO: Consider changing DArray b ->. () to something more general
+
+instance Data.Functor Array where
+  fmap f (Array k n) = Array (\g dest -> k (g . f) dest) n
+
+instance Prelude.Semigroup (Array a) where
+  a <> b = append a b
+instance Semigroup (Array a) where
+  (<>) = append
+
+-- XXX: the use of Vector in the type of alloc is temporary (see also "Data.Array.Destination")
+alloc :: Array a ->. Vector a
+alloc (Array k n) = DArray.alloc n (k id)
+
+walk :: Vector a -> Array a
+walk as =
+  Array
+    (\g dest -> DArray.mirror as g dest)
+    (Vector.length as)
+
+make :: a -> Int -> Array a
+make x n = Array (\k -> DArray.replicate (k x)) n
```

Comment:
Shouldn't we use the `fromFunction` thing instead?

Reply:
I'm not really sure what you mean here - fromFunction right now is for PullArrays not push arrays right now. I guess we could define `make` by creating a pull array and `transfer`ring?

Reply:
But there is one in destination, right? And it would make, I assume, for a better `make` primitive to take an `(Int -> a)`, instead of the `a`. That's what I was, admittedly clumsily, attempting to say.

Reply:
I see! It would, and this would just be implemented by making a PullArray and transferring, I'll make an issue for this.